README.md
koa-limit [![Build Status](https://travis-ci.org/koajs/koa-limit.png)](https://travis-ci.org/koajs/koa-limit)
=========

koa middleware for limit request by ip

## Install

```
npm install koa-limit
```

## Usage

```js
var koa = require('koa');
var favicon = require('koa-favicon');
var limit = require('./');

var app = koa();
// If you are using reverse proxy on the front of node, like 'nginx', please set this
// app.proxy = true;
app.use(favicon());
app.use(limit({
  limit: 1000,
  interval: 1000 * 60 * 60
}));

app.use(function *() {
  this.body = 'hello';
});

app.listen(7001);
```

### Options

* **limit**: limit request number, default is 1000.
* **interval**: limit refresh interval in ms, default is one day.
* **token**: prefix key, for isolate different koa-limit, default is `koa-limit`, if you want to use more than one `koa-limit` in a project, you must set different token!!
* **store**: can be set with a redis store, or other store with API in [MemoryStore](https://github.com/koajs/koa-limit/blob/master/lib/memory_store.js). default store is `MemoryStore`.
* **whiteList**: all ips in whiteList won't be limited.
* **blackList**: all ips in blackList will 403.
* **message**: forbidden message, defautl is 'request frequency limited'.

### Redis Store

checkout the [example.js](https://github.com/koajs/koa-limit/blob/master/example.js), see how to use `koa-limit` with redis.

### Different with [koa-ratelimit](https://github.com/koajs/ratelimit)

1. you can use memory to store all these rates.
2. support `whiteList` and `blackList`
3. easy to store in other DBs.
4. message customable

## Authors

```
$ git summary

 project  : koa-limit
 repo age : 2 days ago
 commits  : 21
 active   : 2 days
 files    : 15
 authors  :
    19  dead_horse              90.5%
     1  Po-Ying Chen            4.8%
     1  fengmk2                 4.8%
```

## License

The MIT License (MIT)

Copyright (c) 2014 dead_horse

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.%


History.md

1.0.2 / 2014-05-10
==================

  * fix retry-after header into second

1.0.1 / 2014-05-10
==================

  * fix black list status error, fix reset header, fixed #3
  * update authors

1.0.0 / 2014-03-13
==================

  * add test for ipchecker
  * Merge pull request #2 from poying/limit
  * ipchecker
  * fix readme

0.0.4 / 2014-03-13
==================

  * fix test
  * add http header, change status code to 429

0.0.3 / 2014-03-13
==================

  * Merge pull request #1 from fengmk2/custom-message
  * let forbidden message customable

0.0.2 / 2014-03-13
==================

  * fix nan

0.0.1 / 2014-03-13
==================

  * complete readme
  * add travis-ci
  * add more test
  * add test
  * add token to make multi limit distinct
  * add frequency limit by ip
  * Initial commit


test/limit.test.js
/*!
 * koa-limit - test/limit.test.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var limit = require('../');
var koa = require('koa');
var mm = require('mm');
var http = require('http');
var request = require('supertest');
var redis = require('redis');
var wrapper = require('co-redis');
var pedding = require('pedding');
var redisClient = require('redis').createClient();
var client = wrapper(redisClient);


function *hello() {
  this.body = 'hello';
}

var appRedis = koa();

appRedis.use(limit({
  token: 'koa-limit:' + process.pid,
  store: client,
  interval: 2000,
  limit: 1
}));

appRedis.use(hello);
appRedis = http.createServer(appRedis.callback());

var appBlack = koa();
appBlack.use(limit({
  blackList: ['127.0.0.*'],
  message: 'access forbidden, please concat foo@bar.com'
}));
appBlack.use(hello);
appBlack = http.createServer(appBlack.callback());

var appWhite = koa();
appWhite.use(limit({
  whiteList: ['127.0.*.*'],
  limit: 0
}));
appWhite.use(hello);
appWhite = http.createServer(appWhite.callback());

describe('test/limit.test.js', function () {
  afterEach(mm.restore);
  describe('blacklist', function () {
    it('should request blackList 403', function (done) {
      request(appBlack)
      .get('/')
      .expect('access forbidden, please concat foo@bar.com')
      .expect(403, done);
    });
  });

  describe('whiteList', function () {
    it('should request whit blackList 200', function (done) {
      request(appWhite)
      .get('/')
      .expect(200, done);
    });
  });

  describe('with redis', function () {

    it('should request 200 but redis record error', function (done) {
      mm(client, 'hincrby', function *() {
        throw new Error('mock error');
      });
      request(appRedis)
      .get('/')
      .expect(200, done);
    });

    it('should request 200', function (done) {
      request(appRedis)
      .get('/')
      .expect('X-RateLimit-Limit', '1')
      .expect('X-RateLimit-Remaining', '0')
      .expect('hello')
      .expect(200, done);
    });

    it('should request 429', function (done) {
      request(appRedis)
      .get('/')
      .expect('X-RateLimit-Limit', '1')
      .expect('X-RateLimit-Remaining', '0')
      .expect('request frequency limited')
      .expect(429, done);
    });

    it('should request 200 if redis get error', function (done) {
      mm(client, 'hget', function *() {
        throw new Error('mock error');
      });
      request(appRedis)
      .get('/')
      .expect(200, done);
    });

    it('should request ok after refresh', function (done) {
      setTimeout(function () {
      request(appRedis)
      .get('/')
      .expect('X-RateLimit-Limit', '1')
      .expect('X-RateLimit-Remaining', '0')
      .expect(200, done);
      }, 2000);
    });
  });
});


test/store.test.js
/*!
 * koa-limit - test/store.test.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var Store = require('../lib/store');
var co = require('co');
var should = require('should');
var mm = require('mm');

var store = new Store({
  ttl: 1000,
  token: 'koa-limit'
});

describe('test/store.test.js', function () {
  afterEach(mm.restore);
  describe('record()', function () {
    it('should record ip not exist ok', co(function *() {
      var key = store.genKey();
      yield store.record('127.0.0.1');
      store.currentHashKey.should.equal(key);
      (yield store.get('127.0.0.1')).should.equal(1);
    }));

    it('should record ip again ok', co(function *() {
      yield store.record('127.0.0.1');
      (yield store.get('127.0.0.1')).should.equal(2);
    }));

    it('should record ok when key changed', co(function *() {
      mm(store, 'genKey', function () {
        return 'new-key';
      });
      yield store.record('127.0.0.1');
      store.currentHashKey.should.equal('new-key');
      (yield store.get('127.0.0.1')).should.equal(1);
    }));
  });

  describe('get', function () {
    it('should get not exist ok', co(function *() {
      (yield store.get('127.0.0.2')).should.equal(0);
    }));

    it('should get not exist ok', co(function *() {
      yield store.record('127.0.0.1');
      yield store.record('127.0.0.1');
      yield store.record('127.0.0.1');
      (yield store.get('127.0.0.1')).should.equal(3);
    }));
  });
});


test/memory_store.test.js
/*!
 * koa-limit - test/memory_store.test.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var MemoryStore = require('../lib/memory_store');
var co = require('co');
var should = require('should');

var store = new MemoryStore();
describe('test/lib/memory_store.test.js', function () {
  describe('hget()', function () {
    it('should get 0 if hashkey not exist', co(function *() {
      (yield store.hget('not_exist', 'field')).should.equal(0);
    }));

    it('should get 0 if field not exist', co(function *() {
      yield store.hincrby('exist', 'field', 1);
      (yield store.hget('exist', 'field_not_exist')).should.equal(0);
    }));

    it('should get ok', co(function *() {
      yield store.hincrby('hashkey', 'field', 1);
      (yield store.hget('hashkey', 'field')).should.equal(1);
    }));
  });

  describe('hincrby()', function () {
    it('should hincry not exist hashkey ok', co(function *() {
      yield store.hincrby('incr_key', 'field', 1);
      (yield store.hget('incr_key', 'field')).should.equal(1);
    }));

    it('should hincry not exist field ok', co(function *() {
      yield store.hincrby('incr_key', 'field_1', 1);
      (yield store.hget('incr_key', 'field_1')).should.equal(1);
    }));

    it('should hincry exist field ok', co(function *() {
      yield store.hincrby('incr_key', 'field_1', 1);
      (yield store.hget('incr_key', 'field_1')).should.equal(2);
    }));
  });
});


lib/store.js
/*!
 * koa-limit - lib/store.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var MemoryStore = require('./memory_store');
var debug = require('debug')('koa-limit:store');

/**
 * Expose `Store`
 */
module.exports = Store;

function Store(options) {
  this.client = options.client || new MemoryStore();
  this.ttl = options.ttl;
  this.currentHashKey;
  this.token = options.token;
};

Store.prototype.genKey = function () {
  return this.token + ':' + Math.floor(Date.now() / this.ttl / 1000);
}

Store.prototype.get = function *(ip) {
  var hashKey = this.genKey();
  var times = yield this.client.hget(hashKey, ip);
  return parseInt(times || '0', 10);
};

Store.prototype.record = function *(ip) {
  var hashKey = this.genKey();
  try {
    yield this.client.hincrby(hashKey, ip, 1);

    if (this.currentHashKey !== hashKey) {
      debug('hashKey changed');
      if (this.client.expire) {
        debug('set expire ttl: %d', this.ttl);
        yield this.client.expire(hashKey, this.ttl);
      } else if (this.currentHashKey) {
        debug('delete old hashkey: %s', this.currentHashKey);
        yield this.client.del(this.currentHashKey);
      }
      this.currentHashKey = hashKey;
    }
  } catch (err) {
    console.warn('record error: ', err.message);
  }
};


lib/limit.js
/*!
 * koa-limit - lib/limit.js
 * Copyright(c) 2014 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var debug = require('debug')('koa-limit');
var ipchecker = require('ipchecker');
var Store = require('./store');

var defaultOptions = {
  interval: 1000 * 60 * 60 * 24,
  limit: 1000,
  whiteList: [],
  blackList: [],
  token: 'koa-limit',
  message: 'request frequency limited',
};

/**
 * limit by ip
 * @param {Object} options
 *   - {Number} interval
 *   - {Number} limit
 *   - {Array} whiteList
 *   - {Array} blackList
 *   - {String} token
 *   - {Object} store
 *   - {String} message
 * @return {[type]} [description]
 */

module.exports = function (options) {
  options = options || {};

  // merge with default
  for (var key in defaultOptions) {
    if (!options[key]) options[key] = defaultOptions[key];
  }

  var store = new Store({
    client: options.store,
    ttl: options.interval / 1000,
    token: options.token
  });

  var whiteListMap = ipchecker.map(options.whiteList);
  var blackListMap = ipchecker.map(options.blackList);

  return function *limit(next) {
    var ip = this.ip;

    if (!ip) {
      debug('can not get ip for the request');
      return yield *next;
    }

    // check black list
    if (ipchecker.check(ip, blackListMap)) {
      debug('request ip: %s is in the blackList', ip);
      this.status = 403;
      this.body = options.message;
      return;
    }
    // check white list
    if (ipchecker.check(ip, whiteListMap)) {
      debug('request ip: %s is in the whiteList', ip);
      return yield *next;
    }
    // check frequency
    var times;
    try {
      times = yield store.get(ip);
    } catch (err) {
      console.warn('get from store error: ', err.message);
      return yield *next;
    }
    var now = Date.now();

    var delta = options.interval - (now % options.interval);
    var reset = (now + delta) / 1000 | 0;
    var remaining = Math.max(options.limit - times - 1, 0);

    this.set('X-RateLimit-Limit', options.limit);
    this.set('X-RateLimit-Remaining', remaining);
    this.set('X-RateLimit-Reset', reset);

    debug('ip: %s request times: %d', ip, times);
    if (times >= options.limit) {
      debug('request frequency limited');
      this.set('Retry-After', delta / 1000 | 0);
      this.status = 429;
      this.body = options.message;
      return;
    }

    yield *next;

    // record
    yield store.record(ip);
  }
};
